File: src/doctor.ts
import { getCurrentBranch, getLatestTag, git, isGitRepository } from "./git";
import { readManifest } from "./manifest";
import { tagPatternForPackage } from "./plan";
import type { NormalizedConfig } from "./types";

export interface DoctorResult {
  errors: string[];
  warnings: string[];
  info: string[];
}

export function runDoctor(cwd: string, config: NormalizedConfig): DoctorResult {
  const result: DoctorResult = { errors: [], warnings: [], info: [] };

  if (config.branches.length === 0 || config.branches.some((branch) => !branch.trim())) {
    result.errors.push("Config must define at least one non-empty release branch.");
  }
  if (config.packages.length === 0) {
    result.errors.push("Config must define at least one package.");
  }
  if (result.errors.length > 0) return result;

  if (!isGitRepository(cwd)) {
    result.errors.push("Current directory is not a git repository.");
    return result;
  }

  if (!git(cwd, ["rev-parse", "--verify", "--quiet", "HEAD^{commit}"], true).trim()) {
    result.errors.push("Repository has no resolvable HEAD commit.");
    return result;
  }

  const branch = getCurrentBranch(cwd);
  if (!config.branches.includes(branch)) {
    result.warnings.push(`Current branch '${branch}' is not a configured release branch.`);
  } else {
    result.info.push(`Release branch: ${branch}`);
  }

  for (const pkg of config.packages) {
    const manifest = readManifest(cwd, pkg);
    result.info.push(`${pkg.name}: manifest version ${manifest.version}`);
    const latestTag = getLatestTag(cwd, tagPatternForPackage(config, pkg));
    if (!latestTag) {
      result.warnings.push(`${pkg.name}: no release tag found; first release will use full reachable history.`);
      continue;
    }

    const tagVersion = extractTagVersion(latestTag);
    result.info.push(`${pkg.name}: latest tag ${latestTag}`);
    if (tagVersion && tagVersion !== manifest.version) {
      result.warnings.push(
        `${pkg.name}: manifest version ${manifest.version} differs from latest tag version ${tagVersion}.`,
      );
    }
  }

  if (config.github !== false && config.github.releases && !process.env.GITHUB_TOKEN && !process.env.GH_TOKEN) {
    result.warnings.push("GITHUB_TOKEN or GH_TOKEN is not set; `release` cannot create GitHub releases.");
  }

  return result;
}

function extractTagVersion(tag: string): string | undefined {
  return /(?:^|@)v(\d+\.\d+\.\d+(?:[-+][^\s]+)?)$/.exec(tag)?.[1];
}

File: internal/doctor/doctor.go
// Package doctor runs the read-only repository and configuration health
// checks reported by `hooversion doctor`. Behavior mirrors src/doctor.ts 1:1,
// including check order, early-return groups, and exact message strings. The
// `ok:` / `warning:` / `error:` prefixes are applied by the CLI layer, so the
// messages here carry no prefixes.
//
// Unlike the TS source, which throws on unreadable manifests or git failures,
// those failures surface as the returned error; all advisory findings are
// reported through the DoctorResult slices in TS order.
package doctor

import (
	"bytes"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/openhoo/hooversion/internal/git"
	"github.com/openhoo/hooversion/internal/manifest"
	"github.com/openhoo/hooversion/internal/plan"
	"github.com/openhoo/hooversion/internal/types"
)

// DoctorResult groups the doctor findings exactly like the TS DoctorResult.
type DoctorResult struct {
	Errors   []string
	Warnings []string
	Infos    []string
}

// tagVersionRE mirrors extractTagVersion's /(?:^|@)v(\d+\.\d+\.\d+(?:[-+][^\s]+)?)$/.
var tagVersionRE = regexp.MustCompile(`(?:^|@)v(\d+\.\d+\.\d+(?:[-+][^\s]+)?)$`)

// RunDoctor performs the doctor checks against cwd using the normalized
// config. getenv abstracts process environment lookups (pass os.Getenv in
// production); a nil getenv falls back to os.Getenv.
func RunDoctor(cwd string, config *types.NormalizedConfig, getenv func(string) string) (DoctorResult, error) {
	result := DoctorResult{Errors: []string{}, Warnings: []string{}, Infos: []string{}}
	if getenv == nil {
		getenv = os.Getenv
	}

	hasBlankBranch := false
	for _, branch := range config.Branches {
		if strings.TrimSpace(branch) == "" {
			hasBlankBranch = true
			break
		}
	}
	if len(config.Branches) == 0 || hasBlankBranch {
		result.Errors = append(result.Errors, "Config must define at least one non-empty release branch.")
	}
	if len(config.Packages) == 0 {
		result.Errors = append(result.Errors, "Config must define at least one package.")
	}
	if len(result.Errors) > 0 {
		return result, nil
	}

	if !git.IsGitRepository(cwd) {
		result.Errors = append(result.Errors, "Current directory is not a git repository.")
		return result, nil
	}

	if !hasResolvableHeadCommit(cwd) {
		result.Errors = append(result.Errors, "Repository has no resolvable HEAD commit.")
		return result, nil
	}

	branch, err := git.CurrentBranch(cwd)
	if err != nil {
		return result, err
	}
	configured := false
	for _, candidate := range config.Branches {
		if candidate == branch {
			configured = true
			break
		}
	}
	if !configured {
		result.Warnings = append(result.Warnings,
			"Current branch '"+branch+"' is not a configured release branch.")
	} else {
		result.Infos = append(result.Infos, "Release branch: "+branch)
	}

	for _, pkg := range config.Packages {
		withAbsManifest := pkg
		withAbsManifest.Manifest = filepath.Join(cwd, pkg.Manifest)
		_, version, err := manifest.Read(withAbsManifest)
		if err != nil {
			return result, err
		}
		result.Infos = append(result.Infos, pkg.Name+": manifest version "+version)

		// TS swallows every getLatestTag failure into "no tag"; LatestTag
		// reports both that case and real git failures as errors.
		tag, err := git.LatestTag(cwd, plan.TagPatternFor(config, pkg))
		if err != nil || strings.TrimSpace(tag) == "" {
			result.Warnings = append(result.Warnings,
				pkg.Name+": no release tag found; first release will use full reachable history.")
			continue
		}

		tagVersion := extractTagVersion(tag)
		result.Infos = append(result.Infos, pkg.Name+": latest tag "+tag)
		if tagVersion != "" && tagVersion != version {
			result.Warnings = append(result.Warnings,
				pkg.Name+": manifest version "+version+" differs from latest tag version "+tagVersion+".")
		}
	}

	if config.GitHub.Releases && getenv("GITHUB_TOKEN") == "" && getenv("GH_TOKEN") == "" {
		result.Warnings = append(result.Warnings,
			"GITHUB_TOKEN or GH_TOKEN is not set; `release` cannot create GitHub releases.")
	}

	return result, nil
}

// hasResolvableHeadCommit mirrors the TS
// `!git(cwd, ["rev-parse", "--verify", "--quiet", "HEAD^{commit}"], true).trim()`
// probe: any failure (unborn HEAD, detached-void, missing git) yields empty
// stdout and therefore false.
func hasResolvableHeadCommit(cwd string) bool {
	cmd := exec.Command("git", "rev-parse", "--verify", "--quiet", "HEAD^{commit}")
	cmd.Dir = cwd
	var stdout bytes.Buffer
	cmd.Stdout = &stdout
	_ = cmd.Run()
	return strings.TrimSpace(stdout.String()) != ""
}

// extractTagVersion returns the semver embedded in a release tag, or "" when
// the tag does not carry one (mirrors extractTagVersion).
func extractTagVersion(tag string) string {
	matches := tagVersionRE.FindStringSubmatch(tag)
	if matches == nil {
		return ""
	}
	return matches[1]
}
